### PR TITLE
Feature/develop2 info package filters

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Fix: Make ``graph info`` filters to work on json output too
Docs: Omit

Close https://github.com/conan-io/conan/issues/12533
There is still some larger work pending in this area, but too much for this PR:
- Improve the serialization of the graph, serializating better the depenencies (requires)
- Use the serialized graph instead of the native Conan object as input to .dot and .html graph renderings

### CODE REVIEW

File: conan/cli/formatters/graph/graph.py [hunk 1]
```diff
@@ -99,6 +100,14 @@ def _render_graph(graph, template, template_folder):
...
     graph = result["graph"]
     conan_api = result["conan_api"]
+    field_filter = result["field_filter"]
+    if field_filter and "requires" not in field_filter:
+        field_filter.append("requires")
+    package_filter = result["package_filter"]
+    serial = graph.serialize()
+    # TODO: This is not used, it is necessary to update the renderings to use the serialized graph
+    #  instead of the native graph
+    serial = filter_graph(serial, package_filter, field_filter)
```
Comment (AbrilRBS) [inline conan/cli/formatters/graph/graph.py]: Do we need to save the output of the function if it's not used elsewhere?

**Review by AbrilRBS — APPROVED** [review]
In the original issue (Or elsewhere) you mentioned that it made little sense to add filtering to outputs meant for further machine processing because you can always filter the resulting json by yourself. I'm guessing that by implementing this, we're walking back on that, right?

File: conan/cli/formatters/graph/graph.py (hunk 1 above)
Comment (memsharded) [reply conan/cli/formatters/graph/graph.py]: It is temporary, with the ``TODO``. The idea is that this ``serial`` is the thing that should be fed to the rendering templates, and not the graph object.

**Comment by memsharded** [comment]
> In the original issue (Or elsewhere) you mentioned that it made little sense to add filtering to outputs meant for further machine processing because you can always filter the resulting json by yourself. I'm guessing that by implementing this, we're walking back on that, right?

Yes, we were reconsidering this, and apparently it would match user expectations, if they provide both the json and a filter is because they want it filtered. It also makes sense to be able to trim the output html/dot graph to the packages that you want, so the filter seems also useful for the other formatters. That seems to mean that all the formatters should be consistent with the user input, filtering out whatever the user wants to filter out. Wdyt?

File: conan/cli/formatters/graph/graph.py (hunk 1 above)
Comment (AbrilRBS) [reply conan/cli/formatters/graph/graph.py]: Ohh now the PR comment makes sense, I didn't quite understand what it was in reference to, ty for the clarification!

**Comment by AbrilRBS** [comment]
Yes, that's how I would have guessed all formatters behaved had I not seen the issues related to this, so I think that this change is very useful for those cases :)

File: conan/cli/formatters/graph/graph.py [hunk 2]
```diff
@@ -99,6 +100,14 @@ def _render_graph(graph, template, template_folder):
 def format_graph_html(result):
     graph = result["graph"]
     conan_api = result["conan_api"]
+    field_filter = result["field_filter"]
+    if field_filter and "requires" not in field_filter:
```
Comment (SSE4) [inline conan/cli/formatters/graph/graph.py]: should it check for `field_filter is not None` instead (as empty list may evaluate to False in python)?

**Review by czoido — APPROVED** [review]

File: conan/cli/formatters/graph/graph.py (hunk 2 above)
Comment (memsharded) [reply conan/cli/formatters/graph/graph.py]: Removing this, it doesn't really apply

**Pull request merged**